src/client/Client.cc

```diff
@@ -13706,7 +13706,7 @@ int Client::_getxattr(Inode *in, const char *name, void *value, size_t size,
     goto out;
   }
 
-  if (!strncmp(name, "ceph.", 5)) {
+  if (!strncmp(name, "ceph.", 5) && strncmp(name, "ceph.mirror.", 12)) { //avoid ceph.mirror attrs
```

Comment:
Although this fixes the get/set, I'm pretty sure its easy to forget which xattr are virtual but start with `ceph.` vs non-virtual (I.e., they are actual xattrs) and start with `ceph.`. @mchangir?

Reply:
In short - what's the terminology for virtual exactly? :)

Reply:
Another option is to treat ceph.mirror.* as virtual xattr on the MDS and thereby fetch it through getvxattr calls. @mchangir

Reply:
If we're mixing virtual and non-virtual in a namespace and need to maintain checks like this, I don't think excluding them via manual strncmp is the way. Can we instead maintain two sets like ```vxattr_names``` and ```real_xattr_names```, with some helper functions so we can query specifically against those when appropriate? I'm worried about building other functionality that stores things in the "ceph." namespace, but then needing to update the MDS code to be allowed to access them. So we should really only check for vxattrs we specifically recognize, rather than the whole namespace?

Reply:
This was indeed discussed during the development phase for cephfs mirroring when the xattr was introduced - https://github.com/ceph/ceph/pull/36603#issuecomment-673493234.

Initially, this xattr was proposed to be a part of the inode itself, but a valid concern was the overhead (memory+ondisk) for each inode and therefore the xattr ended up in xattr_map and for that the xattr needed a bit of special treatment.

Reply:
I would suggest that we do not special case this. Have the client call `_getvxattr` and if it gets `ENODATA`, then check the `in->xattrs` below. It's up to the MDS to control whether `ceph.` xattrs are in the `xattr_map`. That is how the code worked before.

Reply:
Pushed 3 approaches. The reviewer disagreed ones can be dropped.

Reply:
> I would suggest that we do not special case this. Have the client call `_getvxattr` and if it gets `ENODATA`, then check the `in->xattrs` below. It's up to the MDS to control whether `ceph.` xattrs are in the `xattr_map`. That is how the code worked before.

This would work, but, just to discuss a bit more about the virtual vs non-virtual thing - so, we have the following cases:

- virtual xattr stored as part of inode
- virtual xattr stored in xattrmap 
- non-virtual xattr stored as part of inode
- non-virtual xattr stored in xattrmap

(3) from the above list isn't possible and (4) is handled appropriately. So, it was always the case that virtual xattrs (xattr names starting with ceph.) always were a part of the inode until ceph.mirror.* ended up in the xattrmap. I think that's the only xattr that's treated specially @joscollin ?

Reply:
`ceph.mirror.dirty_snap_id` is the only xattr, which is not handled by the MDS (treated specially). `ceph.mirror.info` is also not treated equally by the MDS. They are different. They are treated as [allowed](https://github.com/ceph/ceph/blob/main/src/mds/Server.h#L473). Wondering what classification is that.

Reply:
@vshankar Updated the changes.

Reply:
> `ceph.mirror.dirty_snap_id` is the only xattr, which is not handled by the MDS (treated specially). `ceph.mirror.info` is also not treated equally by the MDS. They are different. They are treated as [allowed](https://github.com/ceph/ceph/blob/main/src/mds/Server.h#L473). Wondering what classification is that.

In that case even ceph.mirror.info would have had the same issue (ENODATA on getxattr), isn't it?

Reply:
I haven't ran through it, but yes it should be the same.